shotover-proxy/src/message/mod.rs

```diff
@@ -426,45 +423,113 @@ impl Value {
-                ColType::Date => Value::NULL,
-                ColType::Time => Value::NULL,
-                ColType::Smallint => {
-                    Value::Integer(decode_smallint(actual_bytes).unwrap() as i64, IntSize::I16)
-                }
-                ColType::Tinyint => {
-                    Value::Integer(decode_tinyint(actual_bytes).unwrap() as i64, IntSize::I8)
-                }
-                // TODO: process collection types based on ColTypeOption
-                // (https://github.com/apache/cassandra/blob/trunk/doc/native_protocol_v4.spec#L569)
-                _ => Value::NULL,
+        let cassandra_type = Value::into_cassandra_type(&spec.col_type, data);
+        Value::create_element(cassandra_type)
+    }
+
+    fn into_cassandra_type(col_type: &ColTypeOption, data: &CBytes) -> CassandraType {
+        let wrapper = get_wrapper_fn(&col_type.id);
+        wrapper(data, col_type)
+    }
+
+    fn create_element(element: CassandraType) -> Value {
+        match element {
+            CassandraType::Ascii(a) => Value::Ascii(a),
+            CassandraType::Bigint(b) => Value::Integer(b, IntSize::I64),
+            CassandraType::Blob(b) => Value::Bytes(b.into_vec().into()),
+            CassandraType::Boolean(b) => Value::Boolean(b),
+            CassandraType::Counter(c) => Value::Counter(c),
+            CassandraType::Decimal(d) => {
+                let big_decimal = BigDecimal::new(d.unscaled, d.scale.into());
+                Value::Decimal(big_decimal)
             }
+            CassandraType::Double(d) => Value::Double(d.into()),
+            CassandraType::Float(f) => Value::Float(f.into()),
+            CassandraType::Int(c) => Value::Integer(c as i64, IntSize::I64),
+            CassandraType::Timestamp(t) => Value::Timestamp(t),
+            CassandraType::Uuid(u) => Value::Uuid(u),
+            CassandraType::Varchar(v) => Value::Varchar(v),
+            CassandraType::Varint(v) => Value::Varint(v),
+            CassandraType::Timeuuid(t) => Value::Timeuuid(t),
+            CassandraType::Inet(i) => Value::Inet(i),
+            CassandraType::Date(d) => Value::Date(d),
+            CassandraType::Time(d) => Value::Time(d),
+            CassandraType::Smallint(d) => Value::Integer(d.into(), IntSize::I16),
+            CassandraType::Tinyint(d) => Value::Integer(d.into(), IntSize::I8),
+            CassandraType::List(_) => Value::create_list(element),
+            CassandraType::Map(_) => Value::create_map(element),
+            CassandraType::Set(_) => Value::create_set(element),
+            CassandraType::Udt(_) => Value::create_udt(element),
+            CassandraType::Tuple(_) => Value::create_tuple(element),
+            CassandraType::Null => Value::NULL,
+        }
+    }
+
+    /// Create a `Value::Udt` from a `CassandraType::Udt`
+    fn create_udt(collection: CassandraType) -> Value {
+        if let CassandraType::Udt(udt) = collection {
+            let mut values = BTreeMap::new();
```

Comment:
This can be written as:
![image](https://user-images.githubusercontent.com/5120858/149060003-48ed7bca-6a1a-4437-ab51-6af37f4866a1.png)

Which preallocates the correct capacity.

And same for the other `BtreeMap`/`BTreeSet`s

Reply:
TIL you can use collect on sets and maps....thought it was just Vec